### DISCUSSION TITLE
Croppable-image detection script -- Help optimising

### DISCUSSION THREAD

**Discussion by gaiking-uk** [ImageMagick/ImageMagick#6465] [Digital Image Processing]
### TLDR;
**Can someone please help me optimise the below code...**

```powershell
### Below is the key part of a wider script to check images for 'croppable edges' ###

# Uses IM to check the std dev. of the top, left, bottom ,right edges of an image
# If result == 0, that edge is solid colour and can be auto-cropped
$T = magick.exe "image.jpg" +repage -crop "x5\!+0+0" -format "%[fx:int(standard_deviation*50)]\n" info:
$L = magick.exe "image.jpg" +repage -crop "5\!x+0+0" -format "%[fx:int(standard_deviation*50)]\n" info:
$B = magick.exe "image.jpg" +repage -crop "x5\!+0+%[fx:h-5]" -format "%[fx:int(standard_deviation*50)]\n" info:
$R = magick.exe "image.jpg" +repage -crop "5\!x+%[fx:w-5]+0" -format "%[fx:int(standard_deviation*50)]\n" info:

### NB: There is later code that processes $T, $L, $B, $R outputs and saves to csv, omitted for brevity ###
```
#### System Info
- Lang: **PowerShell**
- O/S:  **Windows 11**

***

#### Context
- I have a PowerShell script that makes multiple calls to ImageMagick to check if an image has any 'solid colour blocks' on their edges -- i.e. the image can be auro-cropped
- A typical example would be an photo that has been 'padded' by added white/black bars along the sides, etc
- The above script checks the standard deviation of the top, left, bottom, right edges of the image
- If the result = 0, then that edge is a single colour and can be cropped/removed

#### Current Performance
- This script makes 4x IM calls per image
- While the above script works OK, I often run it on a whole folder of images
- A folder of 500 images ends up requiring calling / loading ImageMagick 2,000x times

***

***EDIT: Post updated after replies from @snibgo and @fmw42 to try and add more clarity***

**Comment by snibgo** [comment]
Is your platform bash? Please always state your platform.

Your commands also read `image.jpg` four times. One possibility is:
```
declare $( magick.exe \
  "image.jpg" \
  +repage \
  \( -clone 0 \
     -crop "x5\!+0+0" -format "T=%[fx:int(standard_deviation*50)]\n" \
     -write info: \
     +delete \) \
  \( -clone 0 \
     -crop "5\!x+0+0" -format "L=%[fx:int(standard_deviation*50)]\n" \
     -write info: \
     +delete \) \
  \( -clone 0 \
     -crop "x5\!+0+%[fx:h-5]" -format "B=%[fx:int(standard_deviation*50)]\n" \
     -write info: \
     +delete \) \
  \( -clone 0 \
     -crop "5\!x+%[fx:w-5]+0" -format "R=%[fx:int(standard_deviation*50)]\n" \
     -write info: \
     +delete \) \
  NULL: )

echo $T $L $B $R
```

**Reply by gaiking-uk** [reply]
Hi @snibgo -- Thanks for your reply and apologies for not making it clearer (I swear I remember typing 'PowerShell' when I posted the question originally but re-reading it now, it's not mentioned - my bad).

Also, I wasn't *100%* sure what you meant by *platform* but have edited the post to confirm my language (`PowerShell`) and O/S (`Windows`) assuming that's what you meant -- if not though, please let me know and will try and clarify further.

**Comment by fmw42** [comment]
Here is another option in Unix Imagemagick 7.   I crop the image 4 times and then get the std for each image in the command line and write a test of 0 or 1 depending upon whether it passes the std test. This is output to string of four characters.  Then I extract each character into its own variable. You can then use the variable with values of 0 or 1 to decide to save the file or not.

```
std=`magick redhat.jpg -write mpr:img +delete \
\( mpr:img -crop "x5\!+0+0" \) \
\( mpr:img -crop "5\!x+0+0" \) \
\( mpr:img -crop "x5\!+0+%[fx:h-5]" \) \
\( mpr:img -crop "x5\!+0+%[fx:h-5]" \) \
-format "%[fx:s.standard_deviation<0.01?1:0]" \
info:`
echo $std
std1=`echo ${std:0:1}`
std2=`echo ${std:1:1}`
std3=`echo ${std:2:1}`
std4=`echo ${std:3:1}`
echo "std1=$std1; std2=$std2; std3=$std3; std4=$std4;"

```

**Reply by gaiking-uk** [reply]
Thanks @fmw42 -- I appreciate that as snibgo pointed out, I didn't make my O/S clear so apologies for that... I appreciate the unix code, but am running Windows / PowerShell -- can the above be easily converted to work in that environment?

**Comment by snibgo** [comment]
> Also, I wasn't 100% sure what you meant by platform but have edited the post to confirm my language (PowerShell) and O/S (Windows) assuming that's what you meant -- if not though, please let me know and will try and clarify further.

Yes, by "platform" I really mean the shell language, such as bash or Powershell of Windows BAT or whatever. The O/S is not usually relevant, but might be. 

Sorry, I don't use Powershell, and can't advise. But I expect that either suggestion (mine or fmw42's) can be adapted for Powershell.

**Reply by gaiking-uk** [reply]
OK sure, am happy to try and convert your code into PowerShell equivalent (I also do some windows console or .cmd/.bat scripting experience if that helps)...

To help me deconstruct your code, I was wondering if you can please tell me:
1. I understand the `\` character to be either an escape, or line-continuation character -- is that right?
2. What language is your code written in? -- I am unfamiliar with the `declare` command, is this a bash/linux thing?

**Comment by snibgo** [comment]
1. Yes, in bash, backslash \\ is used for both escape and line-continuation. Windows BAT uses caret ^ for this.

2. Yes, `declare` is a bash command. It takes input lines of the form T=123 and L=234, creating environment variables and assigning values. `magick` creates each of those lines.

If you successfully translate this to Powershell, I encourage you to share that here, to help other people.

**Reply by gaiking-uk** [reply]
Hi @snibgo -- thanks for the follow-up (and apologies for delay replying)

**NB:** The only potential wrinkle is _simultaneous assignment/declaration_ (which I think you code uses). I believe PowerShell also supports this syntax but have not used it before, so was looking into this when I saw the reply from ***dceeee*** come in (see reply below)...

**Comment by dceeee** [comment]
This Powershell script will evaluate the thicknesses of the croppable edges, I hope it helps
Windows10 ImageMagick 7.1.0-32

$infile = "trimthis.png"
$result = magick $infile -trim -set old %W+%H%X%Y +repage -format "%[old]+%@" info:
$result = $result.replace('x','+')
$dim = $result.Split("+")
$Top = $dim[3]
$Left = $dim[2]
$Bottom = $dim[1] - $dim[3] - $dim[5]
$Right = $dim[0] - $dim[2] - $dim[4]
Write-Host Top $Top  Left $Left  Bottom $Bottom  Right $Right
pause

**Reply by gaiking-uk** [reply]
Hi @dceeee -- Thanks for this!

Ironically, my very first draft of the script did an 'auto-crop' all ***ALL*** files (however as this was a lossy process I abandoned this approach pretty quickly) however I like the concept of doing an 'auto-crop' and then comparing it to the original image(s) to see if there ***was*** anything that could be cropped from the image 👍🏼

... It probably isn't a huge surprise that after ***CHECKING*** for "croppable images", my script then goes on to ***CROP*** any applicable images, so again in this case I could simply amend the above to:
1. Generate all 'cropped' images in a temporary directory
2. Compare all original images vs. cropped counterparts
3. For any croppable images: ~~Generate cropped images~~ Move already-generated cropped images back to source folder

The main aim was to improve the efficiency of my original script (I have an image library of 80,000+ images to run this on at some point) and TBH I have no idea how "checking 320,000 image edges (80,000 x 4 edges) for solid colours" compares to "generating cropped versions of 80,000 images" however I will try and find some time to implement the above and let you know.

Thanks again for your help -- will try an post (final) update later this week.
